### PR TITLE
chore(share|das|libs/utils): DRY context resetting in the event of an error for metrics

### LINKED ISSUE #2928: metrics: DRY code pattern
We have the following code pattern in several metrics files: 
```
func (m *metrics) observe(ctx context.Context, observeFn func(context.Context)) {
	if m == nil {
		return
	}

	if ctx.Err() != nil {
		ctx = context.Background()
	}

	observeFn(ctx)
}
```

It would be nice if it could be DRY'd as a utility.

Ref: https://github.com/celestiaorg/celestia-node/pull/2863#discussion_r1372842413

### PR DESCRIPTION
**Pull request by cyneptic**
this creates a function in the utils/resetctx.go that takes a context as an input and resets it if there is an error.
it then uses the function to refactor all the instances of the repeating pattern as requested in the issue - fixes #2928

### CODE REVIEW

File: libs/utils/resetctx.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package utils
+
+import "context"
+
+// ResetContextOnError returns the context if there is no error, otherwise returns it after resetting it to background.
```
Comment (renaynay) [inline libs/utils/resetctx.go:5]:
```suggestion
// ResetContextOnError returns a fresh context if the given context has an error. 
```

**Review by renaynay — COMMENTED** [review]
This is fine with me :) thanks @cyneptic 

Do you mind rebasing on main once https://github.com/celestiaorg/celestia-node/pull/2863

is merged and applying the same to the core package?

**Comment by cyneptic** [comment]
> This is fine with me :) thanks @cyneptic
> 
> Do you mind rebasing on main once #2863
> 
> is merged and applying the same to the core package?

Hi, thank you. tried rebasing and was already up-to-date.
as for the core package, I didn't find any instances of the pattern in there. I applied this to all the patterns I found in the project.

**Comment by cyneptic** [comment]
> This is fine with me :) thanks @cyneptic
> 
> Do you mind rebasing on main once #2863
> 
> is merged and applying the same to the core package?

Hello and thank you.

rebased, is merged to main and applied to core.

also added the suggested change to the function comment.

thank you :)

**Review by renaynay — APPROVED** [review]
lgtm, thanks @cyneptic !

**Review by renaynay — COMMENTED** [review]
@cyneptic Can you please lint 🙏🏻

**Comment by cyneptic** [comment]
> @cyneptic Can you please lint 🙏🏻

hey, i fixed the 6 lint issues that didn't pass last time. golanci-lint doesn't return any errors now, hopefully its okay.

btw github says "dismissed renaynay's stale review" im not sure how that happened or what it is to be honest 😄

**Comment by cyneptic** [comment]
so, whats the status on this pr ? do you want me to update it?

**Review by renaynay — APPROVED** [review]
All good with me :)

**Review by ramin — APPROVED** [review]

**Comment by ramin** [comment]
closes #2928

**Pull request merged**